### DISCUSSION TITLE
support rulesets in languages other than C#

### DISCUSSION THREAD

**Discussion by mokurin000** [ppy/osu#13855] [Ideas] [answered]
[feather](https://github.com/feather-rs/feather/) (don't mind it's a Minecraft server) support plugins by WASM and its API crate
So I guess osu! can implement a WASM VM and write some API library to 
support rulesets in languages supported WASM as target.

My ideas may be very imperfect...

\* Sorry for my poor English *

1. Does wasm's performance good enough to write rhythm games?
2. How to generate WASM API?
...

**Comment by frenzibyte** [comment, marked as answer]
As it has been mentioned before, this will high likely not be supported at all, if it's even possible.

**Reply by mokurin000** [reply]
> As it has been mentioned before, this will high likely not be supported at all, if it's even possible.

https://github.com/RyanLamansky/dotnet-webassembly

I hope to reduce the workload in the future with the development of C# ecosystem
If there is a tool to export C# API to WASM easily in the future, the workload will be even less.

At least there don't seem to be many languages with stable support for WASM target, and I guess this discussion might even hang on until osu!lazer is stable

Anyhow, this will greatly reduce the cost for developers who do not know C# to develop custom rulesets,
though this may not happen until much later.

**Reply by peppy** [reply]
This is 100% out of scope and will not happen on this fork. Please develop it as your own project if you want to see it happen.

**Reply by bdach** [reply]
*If* this was to happen, it would be literal years down the line. And I personally don't believe it should happen, because the sheer idea of having something like rust transpiled to wasm interpreted by c# to be a travesty in terms of complexity. It's a literal babel tower.

This project is written in C#. Ruleset developers will just have to come to terms with it. This is not the minecraft community, this project has two core developers and maybe two to three semi-active, part-time unaffiliated collaborators on top of that. There are not nearly enough resources to make this happen.
